core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -57,13 +63,19 @@ class ControllerChannelManager(controllerEpoch: () => Int,
                                threadNamePrefix: Option[String] = None) extends Logging {
   import ControllerChannelManager._
 
+  type MetricNameWithTagData = java.util.HashMap[String, java.util.Set[java.util.Map[String, String]]]
+
   private val metricsGroup = new KafkaMetricsGroup(this.getClass)
+  // Visible for testing
+  private[controller] val gaugeMetricNameWithTag = new MetricNameWithTagData
```

Comment:
we want to ensure that this is not an ever growing map and we are correctly removing entries from it when applicable. same for other map.